### PR TITLE
Support provisioning with DBaaS operator

### PR DESCRIPTION
**Pull request by jianrongzhang89**
This PR implements the data model design proposed in the ADR for Atlas Cluster provisioning.
https://docs.google.com/document/d/1Jk9SJtZytvh4zBOUf7GJHiwOOSzx6QkLXew6mqNCI0Q/edit

It includes the following:

1) Create a free or non-free MongoDB Atlas cluster through MongoDBAtlasSInstance CR.
2) Update an Atlas cluster by updating the corresponding parameters in the MongoDBAtlasSInstance CR.
3) Delete an Atlas Cluster by deleting MongoDBAtlasSInstance CR.

Example of  MongoDBAtlasSInstance CR when a cluster is created successfully:

apiVersion: dbaas.redhat.com/v1alpha1
kind: MongoDBAtlasInstance
metadata:
  name: test-instance
spec:
  inventoryRef:
    name: jianrzha
    namespace: openshift-dbaas-operator
  name: my-cluster-free
  cloudProvider: aws
  cloudRegion: US_EAST_1
  otherInstanceParams:
    projectName: my-atlas-project-free
    instanceSizeName: M0
Status:
  Conditions:
    Last Transition Time:  2022-01-24T18:18:02Z
    Message:               
    Reason:                Ready
    Status:                True
    Type:                  ProvisionReady
    Last Transition Time:  2022-01-24T18:18:02Z
    Message:               Provider Custom Resource status sync completed
    Reason:                Ready
    Status:                True
    Type:                  InstanceReady
  Instance ID:             61eeeceaa3b8887c9da44a64
  Instance Info:
    Connection Strings Standard Srv:  mongodb+srv://my-cluster-free.10lab.mongodb.net
    Instance Size Name:               M0
    Phase:                            ready
    Project ID:                       61d8ac12e1d16f7f6355ba1a
    Project Name:                     my-atlas-project-free
    Provider Name:                    AWS
    Region Name:                      US_EAST_1
  Phase:                              ready

### CODE REVIEW

File: pkg/controller/atlasinstance/atlasinstance_controller.go [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+	if !ok || len(projectName) == 0 {
+		log.Errorf("Missing %v", dbaas.ProjectNameKey)
+		return nil, fmt.Errorf("missing %v", dbaas.ProjectNameKey)
+	}
+	provider := strings.ToUpper(inst.Spec.CloudProvider)
+	if len(provider) == 0 {
+		provider = "AWS"
+		log.Infof("%v is missing, default value of AWS is used", dbaas.CloudProviderKey)
+	}
+	region := inst.Spec.CloudRegion
```
Comment (redhatHameed) [inline pkg/controller/atlasinstance/atlasinstance_controller.go:322]: We are supposed to pass these values hard coded for this release, as per the trial _version. I mean should be the default region/provider/etc?

**Review by redhatHameed — APPROVED** [review]

File: config/manager/manager.yaml [hunk 2]
```diff
@@ -24,6 +24,7 @@ spec:
         - "--leader-elect"
         - "--health-probe-bind-address=:8081"
         - "--metrics-bind-address=127.0.0.1:8080"
+        - "--sync-period-min=180"
```
[the remark this answers is not in the document]
Comment (jianrongzhang89) [reply config/manager/manager.yaml:27]: The default is 180 min (3 hours)

File: pkg/controller/atlasinstance/atlasinstance_controller.go (hunk 1 above)
Comment (jianrongzhang89) [reply pkg/controller/atlasinstance/atlasinstance_controller.go:322]: For the trial release, the user does not need to pass in region, cloud provider and instance size. When these values are missing, the operator sets the default values for them.

File: config/manager/manager.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (redhatHameed) [reply config/manager/manager.yaml:27]: currently, it's in hours: 
https://github.com/RHEcosystemAppEng/mongodb-atlas-kubernetes/blob/main/main.go#L91

[the remark this answers is not in the document]
Comment (jianrongzhang89) [reply config/manager/manager.yaml:27]: It is hard-coded as 3 hours and not configurable currently. Here we change it to use a parameter with the equivalent default.

[the remark this answers is not in the document]
Comment (redhatHameed) [reply config/manager/manager.yaml:27]: change to use as env var instead of command line WDYT?

[the remark this answers is not in the document]
Comment (redhatHameed) [reply config/manager/manager.yaml:27]: verified the pods will be recreated once update the sync-period value from deployment.

File: main.go [hunk 3]
```diff
@@ -234,6 +249,7 @@ type Config struct {
 	WatchedNamespaces    map[string]bool
 	ProbeAddr            string
 	GlobalAPISecret      client.ObjectKey
+	SyncPeriod           time.Duration
```
Comment (redhatHameed) [inline main.go:252]: not sure, why SyncPeriod passed from AtlasClusterReconciler, is this automatically will apply to all reconcilers

File: config/manager/manager.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jianrongzhang89) [reply config/manager/manager.yaml]: Thanks @redhatHameed. I do not think we need to have both env var and commandline. The env variable approach seems more commonly used. I updated the PR to use env var instead.

File: main.go (hunk 3 above)
Comment (jianrongzhang89) [reply main.go:252]: Yes, it is applied to the manager so used for all reconcilers.

File: pkg/controller/atlasinstance/atlasinstance_controller.go [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+
+func (r *MongoDBAtlasInstanceReconciler) Reconcile(cx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = cx
+	log := r.Log.With("MongoDBAtlasInstance", req.NamespacedName)
+	log.Info("Reconciling MongoDBAtlasInstance")
+
+	inst := &dbaas.MongoDBAtlasInstance{}
+	if err := r.Client.Get(cx, req.NamespacedName, inst); err != nil {
+		if apiErrors.IsNotFound(err) {
+			// CR deleted since request queued, child objects getting GC'd, no requeue
```
Comment (redhatHameed) [inline pkg/controller/atlasinstance/atlasinstance_controller.go:96]: How you are detecting the cluster when MongoDBAtlasInstance is deleted?

Comment (jianrongzhang89) [reply pkg/controller/atlasinstance/atlasinstance_controller.go:96]: When MongoDBAtlasInstance CR is deleted, the corresponding atlascluster CR (via owner references) is deleted. The cluster deletion is then done by atlascluster reconciler's Delete() method.

**Comment by redhatHameed** [comment]
@yzhao583 can you create the MongoDB instance using UI, and see if it works as expected thanks.

**Pull request merged**